### DISCUSSION TITLE
How to write test with mocks using actix-web4?

### DISCUSSION THREAD

**Discussion by eratio08** [SeaQL/sea-orm#642] [Q&A] [answered]
Using the example for [axtic4](https://github.com/SeaQL/sea-orm/blob/master/examples/actix4_example/src/main.rs) I was not able to get tests running which will use a `MockDatabase`. It will fail because the `DatabaseConnection` has to be initialized and then cloned into the closure that initializes the `App`. I understand that this seems to be a deliberate decision.

I then tried to use conditional compilation flags on all parts that required database to be copy
```rust
#[actix_web::main]
#[cfg(not(feature = "mock"))]
async fn main() -> std::io::Result<()> {
...
}
```

Running `cargo test --features mock` with 
```toml
# Cargo.toml
[features]
mock = ["entity/mock"]

# entity/Cargo.toml
[features]
mock = ["sea-orm/mock"]
```
leads to
```shell
   Compiling sea-orm v0.7.1
error[E0277]: the trait bound `OpenTransaction: Clone` is not satisfied
  --> /home/el/.cargo/registry/src/github.com-1ecc6299db9ec823/sea-orm-0.7.1/src/database/mock.rs:14:5
   |
11 | #[derive(Debug, Clone)]
   |                 ----- in this derive macro expansion
...
14 |     transaction: Option<OpenTransaction>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `OpenTransaction`
   |
   = note: required because of the requirements on the impl of `Clone` for `std::option::Option<OpenTransaction>`
   = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
error: could not compile `sea-orm` due to previous error

[Process exited 101]
```

So question now: How could an approach look like, that would allow such tests in an `actix-web4` project?

**Comment by billy1624** [comment, marked as answer]
Hey @eratio08, I'm sorry for the late reply. We're working on demonstrating mock testing on our SeaORM examples. Please check the proposal and working PR at:
- https://github.com/SeaQL/sea-orm/issues/830
- https://github.com/SeaQL/sea-orm/pull/890
